### PR TITLE
add (R)inline-r support for ihaskell

### PR DESCRIPTION
**Pull request by GTrunSec**
the PR feacture that is able to intergrated a suitable environment with  https://github.com/tweag/HaskellR  to support R(inline-r) in ihaskell kernel.
- test Ihaskell setting:
```
 iHaskell = jupyter.kernels.iHaskellWith {
   name = "ihaskell-data-env";
   haskellPackages = pkgs.haskell.packages.ghc865;
   packages = import ./overlay/haskell-list.nix {inherit pkgs;};
   Rpackages = p: with p; [ ggplot2 dplyr xts purrr cmaes cubature
                          reshape2
                         ];                                        
   inline-r = true;
 };
```

- test Jupyter notebook
https://github.com/GTrunSec/Jupyter-data-science-environment/blob/master/TestHaskellAndInlineR.ipynb

image

### CODE REVIEW

**Comment by GTrunSec** [comment]
Also, I would like to PR a new kernel for Julia. But I have a litter confused on how to install Julia package with an argument of `packages`

this my nix-shell test. may you give me some idea or code direction? Thank you.

**Comment by GTrunSec** [comment]
@grahamc  check the finally commits for inline-r. If the function works well in your test.  let me know that. Or if you think this PR  not really necessary merge in master branch. feel free to close the PR. thank you.

**Comment by guaraqe** [comment]
Hello. Sorry, I didn't have time to look at this PR last week. Why did you close it?

**Comment by GTrunSec** [comment]
Setting for iJulia
```
 iJulia = jupyter.kernels.iJuliaWith {
   name =  "julia";
   directory = "/home/gtrun/data/master-jupyter/.julia_pkgs";
   extraPackages = p: with p;[   # GZip.jl # Required by DataFrames.jl
     gzip
     zlib
   ];
 };
```
-Julia  test
image

**Comment by GTrunSec** [comment]
> Hello. Sorry, I didn't have time to look at this PR last week. Why did you close it?
Actually, the PR of Inline-r just optional choice that is alternative install R-packages.
if the user wouldn't like to install R-packages with iHaskell-xxx or has a limited `ihaskell-xxx`  choice. 
- if you have more requirements for R package in Ihaskell.
```
   Rpackages = p: with p; [ ggplot2 dplyr xts purrr cmaes cubature
                          reshape2
                         ];  
```
all of Ihaskell packages as follwing:
```
./pkgs/development/haskell-modules/configuration-hackage2nix.yaml:6414:  - ihaskell-basic
./pkgs/development/haskell-modules/configuration-hackage2nix.yaml:6415:  - ihaskell-display
./pkgs/development/haskell-modules/configuration-hackage2nix.yaml:6416:  - ihaskell-inline-r
./pkgs/development/haskell-modules/configuration-hackage2nix.yaml:6417:  - ihaskell-parsec
./pkgs/development/haskell-modules/configuration-hackage2nix.yaml:6418:  - ihaskell-plot
./pkgs/development/haskell-modules/configuration-hackage2nix.yaml:6419:  - ihaskell-rlangqq
./pkgs/development/haskell-modules/configuration-hackage2nix.yaml:6420:  - ihaskell-widgets
```

File: kernels/default.nix [hunk 1]
```diff
@@ -6,6 +6,7 @@
   iRWith = callPackage ./irkernel;
   iPythonWith = callPackage ./ipython;
   iRubyWith = callPackage ./iruby;
+  iJuliaWith = callPackage ./julia;
```
Comment (guaraqe) [inline kernels/default.nix:9]: This doesn't seem to belong to the PR.

File: kernels/julia/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# JupyterWith IJulia
```
Comment (guaraqe) [inline kernels/julia/README.md:1]: This neither.

File: kernels/julia/default.nix [hunk 3]
```diff
@@ -0,0 +1,85 @@
+{ nixpkgs ? import (builtins.fetchTarball "https://github.com/GTrunSec/nixpkgs/tarball/39247f8d04c04b3ee629a1f85aeedd582bf41cac"){}
```
Comment (guaraqe) [inline kernels/julia/default.nix:1]: This neither.

File: kernels/default.nix (hunk 1 above)
Comment (GTrunSec) [reply kernels/default.nix]: it has been reseted commit to R

**Comment by GTrunSec** [comment]
> I am not sure that most of this PR is needed. Ideally, what I think is a better strategy:
> 
> * Add a new parameter `extraRuntimePackages` to the Haskell kernel, that adds packages to `runtimePackages`, making these packages available to the the kernel.
> * Add the necessary R packages to this parameter when `inline-r` is being used.
> 
> If this works, this is way more extensible, since this approach would work for `inline-java` too, for example.
> 
> Can you check that this works?

Does the `inline-java` need to set a specific path such as:
```
 export R_LIBS_SITE=${builtins.readFile r-libs-site}
```
the inline-r default of false. this PR will make the user who needs more extensible R packages instead of necessary R packages.

**Comment by guaraqe** [comment]
I think I didn't make myself very clear. What I propose is that, instead of introducing R specific values, add two parameters:

- `extraRuntimePackages`, which adds the packages to the kernel environment.
- `extraEnvVars`, which adds extra environment variables to the environment.

You have a good part of this done, it shouldn't be hard to change.

In that way, this can be useful for more use-cases, while covering the `inline-r` case. We can add a section to the `IHaskell` README to explain how to setup `inline-r`, and progressively expand to other libraries of the same kind. In that way, we keep the code as simple as possible, avoiding many branches that can fail differently.
